README.md
# `CIPerm`: Computationally-Efficient Confidence Intervals for Mean Shift from Permutation Methods

<!-- badges: start -->
[![R-CMD-check](https://github.com/ColbyStatSvyRsch/CIPerm/workflows/R-CMD-check/badge.svg)](https://github.com/ColbyStatSvyRsch/CIPerm/actions)
[![CRAN downloads](https://cranlogs.r-pkg.org/badges/CIPerm?color=5f810e)](https://cran.r-project.org/package=CIPerm)
<!-- badges: end -->

This R package implements computationally-efficient construction of confidence intervals from permutation tests or randomization tests for simple differences in means. In other words, if we set up a permutation or randomization test to evaluate $H_0: \mu_A - \mu_B = 0$,
then how can we use that same single set of permutations to cheaply construct a CI for the $(\mu_A - \mu_B)$ parameter?

The method is based on Minh D. Nguyen's 2009 MS thesis paper, "Nonparametric Inference using Randomization and Permutation Reference Distribution and their Monte-Carlo Approximation," https://doi.org/10.15760/etd.7798.
See the `nguyen` vignette for a brief summary of the method and for our replication of Nguyen's results.

Note that our R function arguments and outputs are structured differently than the similarly-named R functions in Nguyen (2009), but the results are equivalent.

A copy of our [useR! 2022 conference](https://user2022.r-project.org/) poster about the package is in our GitHub repo, in the [`data-raw` folder](https://github.com/ColbyStatSvyRsch/CIPerm/tree/master/data-raw).

## Installation

Install a stable version of the package from [CRAN](https://cran.r-project.org/package=CIPerm):

```r
install.packages("CIPerm")
```

Or, for the latest development version, install directly from GitHub:

```r
# install.packages("remotes")
remotes::install_github("ColbyStatSvyRsch/CIPerm", build_vignettes = TRUE)
```

## Usage

First use `dset(x,y)` to tabulate summary statistics for each permutation.
Then pass the results into `cint()` to compute a confidence interval.

```r
x <- c(19, 22, 25, 26)
y <- c(23, 33, 40)
demo <- dset(x, y)
cint(dset = demo, conf.level = .95, tail = "Two")
#> [1] -21   3
```

You can also pass the results of `dset()` into `pval()` to calculate p-values
for a difference in means (`value="m"`),
a difference in medians (`value="d"`),
or the Wilcoxon rank sum test (`value="w"`).

```r
pval(dset = demo, tail = "Left", value = "m")
#> [1] 0.08571429
pval(dset = demo, tail = "Left", value = "d")
#> [1] 0.08571429
pval(dset = demo, tail = "Left", value = "w")
#> [1] 0.1142857
```

See also our `naive` vignette for timing comparisons of Nguyen's one-pass method. In the "naive" approach, you would run many separate permutation tests with different null values for $(\mu_A - \mu_B)$, and your CI would consist of those values where the null hypothesis was not rejected. The vignette shows that if you need to check more than a few null values and your dataset isn't trivially small, Nguyen's method can be considerably faster than the naive approach.

## References

Ernst, M.D. (2004).
"Permutation Methods: A Basis for Exact Inference,"
*Statistical Science*, vol. 19, no. 4, 676-685,
[DOI:10.1214/088342304000000396](https://doi.org/10.1214/088342304000000396).

Nguyen, M.D. (2009).
"Nonparametric Inference using Randomization and Permutation
Reference Distribution and their Monte-Carlo Approximation"
[unpublished MS thesis; Mara Tableman, advisor], Portland State University.
*Dissertations and Theses*. Paper 5927.
[DOI:10.15760/etd.7798](https://doi.org/10.15760/etd.7798).

Tupaj, E. and Wieczorek, J. (2022).
"`CIPerm`: An R Package for Computationally Efficient Confidence Intervals from Permutation Test,"
poster presented at the useR! Conference, held virtually, June 22, 2022.
https://user2022.r-project.org/program/posters/


NEWS.md
# CIPerm 0.2.3

* Fix typo in URLs.

# CIPerm 0.2.2

* Update Nguyen (2009) links from URLs to DOIs.
* Add package URL and BugReports to DESCRIPTION.
* Add `naive` vignette, which compares computation speeds of Nguyen's method vs a "naive" grid search.
* Add useR! 2022 conference poster to `data-raw` folder. Cite conference poster in README.

# CIPerm 0.2.1

* Add link for Nguyen (2009) citation.
* Show function arguments in examples and vignette.

# CIPerm 0.2.0

* Initial package pre-release.


R/cint.R
#' Permutation-methods confidence interval for difference in means
#'
#' Calculate confidence interval for a simple difference in means
#' from a two-sample permutation or randomization test.
#' In other words, we set up a permutation or randomization test to evaluate
#' \eqn{H_0: \mu_A - \mu_B = 0}, then use those same permutations to
#' construct a CI for the parameter \eqn{\delta = (\mu_A - \mu_B)}.
#'
#' If the desired \code{conf.level} is not exactly feasible,
#' the achieved confidence level will be slightly anti-conservative.
#' We use the default numeric tolerance in \code{\link{all.equal}} to check
#' if \code{(1-conf.level) * nrow(dset)} is an integer for one-tailed CIs,
#' or if \code{(1-conf.level)/2 * nrow(dset)} is an integer for two-tailed CIs.
#' If so, \code{conf.level.achieved} will be the desired \code{conf.level}.
#' Otherwise, we will use the next feasible integer,
#' thus slightly reducing the confidence level.
#' For example, in the example below the randomization test has 35 combinations,
#' and a two-sided CI must have at least one combination value in each tail,
#' so the largest feasible confidence level for a two-sided CI is 1-(2/35) or around 94.3\%.
#' If we request a 95\% or 99\% CI, we will have to settle for a 94.3\% CI instead.
#'
#' @param dset The output of \code{\link{dset}}.
#' @param conf.level Confidence level (default 0.95 corresponds to 95\% confidence level).
#' @param tail Which tail? Either "Two"- or "Left"- or "Right"-tailed interval.
#' @return A list containing the following components:\describe{
#'   \item{\code{conf.int}}{Numeric vector with the CI's two endpoints.}
#'   \item{\code{conf.level.achieved}}{Numeric value of the achieved confidence level.}
#' }
#' @examples
#' x <- c(19, 22, 25, 26)
#' y <- c(23, 33, 40)
#' demo <- dset(x, y)
#' cint(dset = demo, conf.level = .95, tail = "Two")
#' @export


cint <- function(dset, conf.level = .95, tail = c("Two", "Left", "Right")){

  sig <- 1 - conf.level
  stopifnot(sig > 0 & sig < 1)
  tail <- match.arg(tail)

  num <- nrow(dset)

  w.i <- sort(dset$wkd, decreasing = FALSE, na.last = FALSE)

  # Keeping an eye on how we use w.i later...
  #   When dset's nmc leads us to use Monte Carlo sims,
  #   we may get some permutations equivalent to orig data
  #   i.e. we may get SEVERAL k=0 and therefore several w.i=NaN.
  # For this reason, some hardcoded "1"s below
  #   (which used to be correct when we didn't have a Monte Carlo option)
  #   must be replaced with the number of k=0 rows.
  nk0 <- sum(dset$k == 0)
  stopifnot(nk0 >= 1) # Our code assumes 1st row is orig data, so k=0 at least once

  # Below we'll need to take ceiling(siglevel*num) several times.
  # BUT this has caused problems when siglevel*num "should be" an integer
  # but was represented as sliiightly more than that integer in floating point...
  # So here is an internal function which will EITHER
  # 1. use round(siglevel*num), if siglevel*num is basically an integer,
  #    to within the default numerical tolerance of all.equal();
  # 2. or otherwise use ceiling(siglevel*num) instead.
  roundOrCeiling <- function(x) {
    ifelse(isTRUE(all.equal(round(x), x)), # is x==round(x) to numerical tolerance?
           round(x),
           ceiling(x))
  }
  ## Quick checks:
  ## ceiling() fails to give 2, but roundOrCeiling() works correctly
  # > ceiling((1-(1-2/35))*35) ## whoops!
  # [1] 3
  # > roundOrCeiling((1-(1-2/35))*35)
  # [1] 2
  ## Another check:
  ## ceiling() fails to give 250, but roundOrCeiling() works correctly
  # > ceiling((1-.95)*5000) ## whoops!
  # [1] 251
  # > roundOrCeiling((1-.95)*5000)
  # [1] 250


  if (tail == "Left"){
    siglevel <- sig
    index <- roundOrCeiling(siglevel*num) - 1
    UB <- w.i[(num-index)]
    LT <- c(-Inf, UB)
    conf.achieved <- 1-((index+1)/num)
    message(paste0("Achieved conf. level: 1-(", index+1, "/", num, ")"))
    return(list(conf.int = LT,
                conf.level.achieved = conf.achieved))
  } else if (tail == "Right"){
    siglevel <- sig
    index <- roundOrCeiling(siglevel*num) - 1
    LB <- w.i[1+nk0+index] # starts counting from the (1+nk0)'th element of w.i
    # (not the first (original) which will always be 'NaN')
    RT <- c(LB, Inf)
    conf.achieved <- 1-((index+1)/num)
    message(paste0("Achieved conf. level: 1-(", index+1, "/", num, ")"))
    return(list(conf.int = RT,
                conf.level.achieved = conf.achieved))
  } else { # tail == "Two"
    siglevel <- sig/2  # use half of sig in each tail
    index <- roundOrCeiling(siglevel*num) - 1
    UB <- w.i[(num-index)]
    LB <- w.i[1+nk0+index] # starts counting from the (1+nk0)'th element of w.i
    # (not the first (original) which will always be 'NaN')
    Upper <- if(is.na(UB)) Inf else UB
    Lower <- if(is.na(LB)) -Inf else LB
    CI <- c(Lower, Upper)
    conf.achieved <- 1-(2*(index+1)/num)
    message(paste0("Achieved conf. level: 1-2*(", index+1, "/", num, ")"))
    return(list(conf.int = CI,
                conf.level.achieved = conf.achieved))
  }

}


R/dset.R
#' Permutation-methods summary statistics
#'
#' Calculate table of differences in means, medians, etc. for each
#' combination (or permutation, if using Monte Carlo approx.),
#' as needed in order to compute a confidence interval using \code{\link{cint}}
#' and/or a p-value using \code{\link{pval}}.
#'
#' @param group1 Vector of numeric values for first group.
#' @param group2 Vector of numeric values for second group.
#' @param nmc Threshold for whether to use Monte Carlo draws or complete
#'   enumeration. If the number of all possible combinations
#'   \code{choose(n1+n2, n1) <= nmc}, we use complete enumeration.
#'   Otherwise, we take a Monte Carlo sample of \code{nmc} permutations.
#'   You can set \code{nmc = 0} to force complete enumeration regardless of
#'   how many combinations there are.
#' @param returnData Whether the returned dataframe should include columns for
#'   the permuted data itself (if TRUE), or only the derived columns that are
#'   needed for confidence intervals and p-values (if FALSE, default).
#' @return A data frame ready to be used in \code{cint()} or \code{pval()}.
#' @examples
#' x <- c(19, 22, 25, 26)
#' y <- c(23, 33, 40)
#' demo <- dset(x, y, returnData = TRUE)
#' knitr::kable(demo, digits = 2)
#' @export


######################
#### Old notes-to-self from when we used to have two versions,
####   dset() using for-loops vs a vectorized dsetFast().
####   As of 3/22/2022, dsetFast() became the current version of dset() below,
####   and the old version was simply removed.
####   Look on GitHub for package versions 0.1.0.9004 or earlier
####   (March 21, 2022 or earlier) to see old dset() code.
##
## CHECK TIMINGS AND OUTPUTS OF dset() VS dsetFast():
##
## Skipping the for loop with dsetFast() gives a ~7x speedup over original dset()!
# > x <- c(19, 22, 25, 26)
# > y <- c(23, 33, 40)
# > microbenchmark::microbenchmark(dset(x, y), dsetFast(x, y), times = 1000)
# Unit: microseconds
#           expr      min       lq     mean   median       uq       max neval
#     dset(x, y) 6276.401 6790.501 7715.261 7210.501 7910.150 51722.501  1000
# dsetFast(x, y)  901.101 1060.551 1215.499 1141.201 1238.651  4548.201  1000
##
## And it does give the same result!
##
# > demo <- dset(x, y)
# > demo2 <- dsetFast(x, y)
# > all.equal(demo, demo2)
# [1] TRUE
##
## Let's check this on other datasets too,
## and let's check Monte Carlo runs with set.seed()...
##
## What about the larger groups of 6+9 obs instead of 4+3 obs?
# > wl1 <- c(1.72, 1.64, 1.74, 1.70, 1.82, 1.82, 1.90, 1.82, 2.08)
# > wl2 <- c(1.78, 1.86, 1.96, 2.00, 2.00, 1.96)
# > microbenchmark::microbenchmark(dset(wl1, wl2), dsetFast(wl1, wl2), times = 50)
# Unit: milliseconds
#               expr       min       lq       mean     median        uq       max neval
#     dset(wl1, wl2) 1138.0176 1226.140 1278.69760 1278.08755 1327.2168 1438.2427    50
# dsetFast(wl1, wl2)   43.3406   47.194   50.46918   49.85155   52.8262   77.2373    50
##
## OH WOW, this is 25x faster now.
## Here, plain dset() takes ~1sec each time, while dsetFast() is ~0.05sec.
## I imagine the diffs will only get more extreme for larger datasets.
##
## And they still give same results?
# > demo <- dset(wl1, wl2)
# > demo2 <- dsetFast(wl1, wl2)
# > all.equal(demo, demo2)
# [1] TRUE
# > demo <- dset(wl1, wl2, returnData = TRUE)
# > demo2 <- dsetFast(wl1, wl2, returnData = TRUE)
# > all.equal(demo, demo2)
# [1] TRUE
## Yes they do!
##
## What about Monte Carlo runs?
## They shouldn't match without set.seed(), and indeed they don't...
# > demo <- dset(wl1, wl2, nmc = 5000)
# > demo2 <- dsetFast(wl1, wl2, nmc = 5000)
# > isTRUE(all.equal(demo, demo2))
# [1] FALSE
##
## But WITH set.seed, they DO match, as they should!
# > seed = 20220321
# > set.seed(seed)
# > demo <- dset(wl1, wl2, nmc = 5000)
# > set.seed(seed)
# > demo2 <- dsetFast(wl1, wl2, nmc = 5000)
# > all.equal(demo, demo2)
# [1] TRUE
##
## So old dset() can be replaced with new dsetFast()
######################



# TODO: in the future, consider switching to RcppAlgos::comboSample(),
# which lets you sample *directly* from the set of all possible combinations
# without having to generate them all first.
# However, that would force us to Require several other packages.
# For now, let's start with just using base sample()
# (even though it gives permutations, not only unique combinations,
#  so that we might get multiple perms equivalent to the same combo;
#  we'll need to handle multiple perms equivalent to original data grouping).


dset <- function(group1, group2, nmc = 10000, returnData = FALSE){
  stopifnot(nmc >= 0)
  stopifnot(nmc != 1)
  stopifnot(is.numeric(group1) & is.numeric(group2))
  stopifnot(length(group1) >= 1 & length(group2) >= 1)
  stopifnot(!any(is.na(c(group1, group2))))

  combined <- c(group1, group2)

  n <- length(group1)
  m <- length(group2)
  den <- (1/n + 1/m)

  N <- n + m
  num <- choose(N, n) # number of possible combinations

  # Form a matrix where each column contains indices in new "group1" for that comb or perm
  if(nmc == 0 | num <= nmc) { # take all possible combinations
    dcombn1 <- utils::combn(1:N, n)
  } else { # use Monte Carlo sample of permutations, not all possible combinations
    dcombn1 <- replicate(nmc, sample(N, n))
    dcombn1[,1] <- 1:n # force the 1st "combination" to be original data order
    num <- nmc
  }

  # Form the equivalent matrix for indices in new "group2"
  dcombn2 <- apply(dcombn1, 2, function(x) setdiff(1:N, x))

  # Form the corresponding matrices of data values, not data indices
  group1_perm <- matrix(combined[dcombn1], nrow = n)
  group2_perm <- matrix(combined[dcombn2], nrow = m)

  # For each comb or perm, compute:
  #   difference in group means; sum in group1; difference in group medians;
  #   and sum of *ranks* in group1 (the statistic for the Wilcoxon rank sum test)
  diffmean <- colMeans(group1_perm) - colMeans(group2_perm)
  sum1 <- colSums(group1_perm)
  diffmedian <- matrixStats::colMedians(group1_perm) - matrixStats::colMedians(group2_perm)
  r <- rank(combined, ties.method = "first")
  wilsum <- colSums(matrix(r[dcombn1], nrow = n))

  # For each comb or perm, compute:
  #   k = how many values swapped from group1 to group2?
  #   wkd = Nguyen (2009) statistic whose quantiles are used for CI endpoints
  k <- colSums(matrix(dcombn1 %in% ((n+1):N), nrow=n))
  wkd <- (diffmean[1] - diffmean) / (k * den)

  dataframe <- data.frame(diffmean = diffmean,
                          sum1 = sum1,
                          diffmedian = diffmedian,
                          wilsum = wilsum,
                          k = k,
                          wkd = wkd)

  if(returnData){
    # Return the whole table formatted as in Nguyen (2009)
    # for comparison with his results
    w.i <- sort(dataframe$wkd, decreasing = FALSE, na.last = FALSE)
    ID <- c(1:num)
    dataset <- t(rbind(group1_perm, group2_perm))
    datatable <- cbind(ID, dataset, dataframe, w.i)
    return(datatable)
  } else {
    # Just return the fewest columns needed for p-vals and CIs
    return(dataframe)
  }

}


R/pval.R
#' Permutations-methods p-values for difference in means, medians, or Wilcoxon rank sum test
#'
#' Calculate p-values for a two-sample permutation or randomization test.
#' In other words, we set up a permutation or randomization test to evaluate
#' the null hypothesis that groups A and B have the same distribution,
#' then calculate p-values for several alternatives:
#' a difference in means (\code{value="m"}),
#' a difference in medians (\code{value="d"}),
#' or the Wilcoxon rank sum test (\code{value="w"}).
#'
#' @param dset The output of \code{\link{dset}}.
#' @param tail Which tail? Either "Two"- or "Left"- or "Right"-tailed test.
#' @param value Either "m" for difference in means (default);
#'     "s" for sum of Group 1 values
#'     [equivalent to "m" and included only for sake of checking results against
#'     Nguyen (2009) and Ernst (2004)];
#'     "d" for difference in medians;
#'     or "w" for Wilcoxon rank sum statistic;
#'     or "a" for a named vector of all four p-values.
#' @return Numeric p-value for the selected type of test,
#'     or a named vector of all four p-values if \code{value="a"}.
#' @examples
#' x <- c(19, 22, 25, 26)
#' y <- c(23, 33, 40)
#' demo <- dset(x, y)
#' pval(dset = demo, tail = "Left", value = "s")
#' pval(dset = demo, tail = "Left", value = "a")
#' @export


# TODO: Should two-tailed tests be for |x-median| and not |x-mean| ???
#   The p-value involves finding how much of null distr
#   is further away from the 50th percentile, right?...
# However, Ernst (2004) paper in *Statistical Science* uses |x-mean|,
#   so let's stick with that for now.



pval <- function(dset, tail = c("Two", "Left", "Right"),
                 value = c("m", "s", "d", "w", "a")){

  tail <- match.arg(tail)
  value <- match.arg(value)

  num <- nrow(dset)

  if (tail == "Left"){
    pvalmean <- sum(dset$diffmean <= dset$diffmean[1])/num
    pvalsum <- sum(dset$sum1 <= dset$sum1[1])/num
    pvalmedian <- sum(dset$diffmedian <= dset$diffmedian[1])/num
    pvalwilsum <- sum(dset$wilsum <= dset$wilsum[1])/num
  } else if (tail == "Right") {
    pvalmean <- sum(dset$diffmean >= dset$diffmean[1])/num
    pvalsum <- sum(dset$sum1 >= dset$sum1[1])/num
    pvalmedian <- sum(dset$diffmedian >= dset$diffmedian[1])/num
    pvalwilsum <- sum(dset$wilsum >= dset$wilsum[1])/num
  } else { # tail == "Two"
    pvalmean <- sum(abs(dset$diffmean - mean(dset$diffmean)) >= abs(dset$diffmean[1] - mean(dset$diffmean)))/num
    pvalsum <- sum(abs(dset$sum1 - mean(dset$sum1)) >= abs(dset$sum1[1] - mean(dset$sum1)))/num
    pvalmedian <- sum(abs(dset$diffmedian - mean(dset$diffmedian)) >= abs(dset$diffmedian[1] - mean(dset$diffmedian)))/num
    pvalwilsum <- sum(abs(dset$wilsum - mean(dset$wilsum)) >= abs(dset$wilsum[1] - mean(dset$wilsum)))/num
  }

  if (value == "m"){
    return(pvalmean)
  } else if (value == "s"){
    return(pvalsum)
  } else if (value == "d"){
    return(pvalmedian)
  } else if (value == "w"){
    return(pvalwilsum)
  } else { # value == "a"
    return(c(m = pvalmean,
             s = pvalsum,
             d = pvalmedian,
             w = pvalwilsum))
  }

}


R/CIPerm.R
#' CIPerm: Computationally-Efficient Confidence Intervals for Mean Shift from Permutation Methods
#'
#' Implements computationally-efficient construction of
#' confidence intervals from permutation tests or randomization tests
#' for simple differences in means.
#' The method is based on Minh D. Nguyen's 2009 MS thesis paper,
#' "Nonparametric Inference using Randomization and Permutation
#' Reference Distribution and their Monte-Carlo Approximation,"
#' <\doi{10.15760/etd.7798}>
#' See the \code{nguyen} vignette for a brief summary of the method.
#' First use \code{\link{dset}} to tabulate summary statistics for each permutation.
#' Then pass the results into \code{\link{cint}} to compute a confidence interval,
#' or into \code{\link{pval}} to calculate p-values.
#'
#' Our R function arguments and outputs are structured differently
#' than the similarly-named R functions in Nguyen (2009),
#' but the results are equivalent. In the \code{nguyen} vignette
#' we use our functions to replicate Nguyen's results.
#'
#' Following Ernst (2004) and Nguyen (2009), we use "permutation methods"
#' to include both randomization tests and permutation tests.
#' In the simple settings in this R package,
#' the randomization and permutation test mechanics are identical,
#' but their interpretations may differ.
#'
#' We say "randomization test" under the model where
#' the units are not necessarily a random sample, but the treatment assignment
#' was random. The null hypothesis is that the treatment has no effect.
#' In this case we can make causal inferences about the
#' treatment effect (difference between groups) for this set of individuals,
#' but cannot necessarily generalize to other populations.
#'
#' By contrast, we say "permutation test" under the model where
#' the units were randomly sampled from two distinct subpopulations.
#' The null hypothesis is that the two groups have identical CDFs.
#' In this case we can make inferences about differences between subpopulations,
#' but there's not necessarily any "treatment" to speak of
#' and causal inferences may not be relevant.
#'
#' @references Ernst, M.D. (2004).
#'   "Permutation Methods: A Basis for Exact Inference,"
#'   \emph{Statistical Science}, vol. 19, no. 4, 676-685,
#'   <\doi{10.1214/088342304000000396}>.
#'
#'   Nguyen, M.D. (2009).
#'   "Nonparametric Inference using Randomization and Permutation
#'   Reference Distribution and their Monte-Carlo Approximation"
#'   [unpublished MS thesis; Mara Tableman, advisor], Portland State University.
#'   \emph{Dissertations and Theses}. Paper 5927.
#'   <\doi{10.15760/etd.7798}>.


#' @importFrom matrixStats colMedians
#' @importFrom stats median
#' @importFrom utils combn
#'
#' @docType package
#' @name CIPerm
NULL


